src/core/memory.d

```diff
@@ -919,3 +919,186 @@ extern (C) private pure @system @nogc nothrow
 
     pragma(mangle, "free") void fakePureFree(void* ptr);
 }
+
+private extern (C) void _d_delinterface(void** p);
+private extern (C) void _d_delclass(Object* p);
+private extern (C) void _d_delstruct(void** p, TypeInfo_Struct inf);
+
+/**
+The `delete` keyword allowed to free GC-allocated memory.
+As this is inherently not `@safe`, it has been deprecated.
+This function has been added to provide an easy transition from `delete`.
+It performs the same functionality as `delete`.
+
+Note: Users should prefer $(REF destroy, object)` to explicitly finalize objects,
+and only resort to $(REF __delete, core,memory) when $(REF destroy, object)
+would not be a feasible option.
+
+In detail, `__delete` does the following:
+$(UL
+
+    $(LI If `t` is a class object reference, and
+        there is a destructor for that class, the destructor
+        is called for that object instance.
+    )
+    $(LI Otherwise, the garbage collector is called to immediately free the
+        memory allocated for the class instance or struct instance.
+        If the garbage collector was not used to allocate the memory for
+        the instance, undefined behavior will result.
+    )
+    $(LI If `t` is a pointer or a dynamic array,
+        the garbage collector is called to immediately release the
+        memory.
+        If the garbage collector was not used to allocate the memory for
+        the instance, undefined behavior will result.
+    )
+    $(LI The pointer, dynamic array, or reference is set to `null`
+        after the delete is performed.
+        Any attempt to reference the data after the deletion via another
+        reference to it will result in undefined behavior.
+    )
+)
+
+Params:
+    t = aggregate object that should be destroyed
+
+See_Also: $(REF destroy, object), $(REF free, core,GC)
+*/
+void __delete(T)(ref T t) @system
+{
+    static if (is(T == interface))
+    {
+        _d_delinterface(cast(void**) &t);
+    }
+    else static if (is(T == class))
+    {
+        _d_delclass(cast(Object*) &t);
+    }
+    else static if (is(T == struct))
+    {
+        _d_delstruct(cast(void**) &t, typeid(T));
```

Comment:
Wait, this is incorrect. `__delete` should be only applicable to pointers to struct. Let's not innovate here.